mmdeploy/codebase/mmdet3d/models/centerpoint.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmdet3d.core import circle_nms
+
+from mmdeploy.core import FUNCTION_REWRITER
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmdet3d.models.detectors.centerpoint.CenterPoint.extract_pts_feat')
+def centerpoint__extract_pts_feat(ctx, self, voxels, num_points, coors,
+                                  img_feats, img_metas):
+    """Extract features from points. Rewrite this func to remove voxelize op.
+
+    Args:
+        voxels (torch.Tensor): Point features or raw points in shape (N, M, C).
+        num_points (torch.Tensor): Number of points in each voxel.
+        coors (torch.Tensor): Coordinates of each voxel.
+        img_feats (list[torch.Tensor], optional): Image features used for
+            multi-modality fusion. Defaults to None.
+        img_metas (list[dict]): Meta information of samples.
+
+    Returns:
+        torch.Tensor: Points feature.
+    """
+    if not self.with_pts_bbox:
+        return None
+
+    voxel_features = self.pts_voxel_encoder(voxels, num_points, coors)
+    batch_size = coors[-1, 0] + 1
+    x = self.pts_middle_encoder(voxel_features, coors, batch_size)
+    x = self.pts_backbone(x)
+    if self.with_pts_neck:
+        x = self.pts_neck(x)
+    return x
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmdet3d.models.detectors.centerpoint.CenterPoint.simple_test_pts')
+def centerpoint__simple_test_pts(ctx, self, x, img_metas, rescale=False):
+    """Rewrite this func to format model outputs.
+
+    Args:
+        x (torch.Tensor): Input points feature.
+        img_metas (list[dict]): Meta information of samples.
+        rescale (bool): Whether need rescale.
+
+    Returns:
+        List: Result of model.
+    """
+    outs = self.pts_bbox_head(x)
+    bbox_preds, scores, dir_scores = [], [], []
+    for task_res in outs:
+        bbox_preds.append(task_res[0]['reg'])
+        bbox_preds.append(task_res[0]['height'])
+        bbox_preds.append(task_res[0]['dim'])
+        if 'vel' in task_res[0].keys():
+            bbox_preds.append(task_res[0]['vel'])
+        scores.append(task_res[0]['heatmap'])
+        dir_scores.append(task_res[0]['rot'])
+    bbox_preds = torch.cat(bbox_preds, dim=1)
+    scores = torch.cat(scores, dim=1)
+    dir_scores = torch.cat(dir_scores, dim=1)
+    return scores, bbox_preds, dir_scores
+
+
+@FUNCTION_REWRITER.register_rewriter(
+    'mmdet3d.models.dense_heads.centerpoint_head.CenterHead.get_bboxes')
+def centerpoint__get_bbox(ctx,
```

Comment:
Better not rewrite a method if it has not been used in export. And we should avoid changing the behavior of the method.

Reply:
I know, but centerpoint_head output num is task_num*N, and task_num is variable. So, I need to rewrite head to format output to fix output num=3 and split head output in post process.

Reply:
Can we add some code before/after invoke get_bboxes to avoid the rewrite?